Add top-level sync command

Nydauron opened the pull request:
# Description
NOTE: The script will break once #480 is merged due to the references to env vars. I'd suggest first merging #480, then fixing `sync_commands.py` and merging this PR.

Adds a sync command and sync Python script to sync bot commands. A script is provided in case the sync command signature ever gets modified. (`/sync` would result in a "The application did not respond" error)


# Checks

- [x] I ran `black` over the code to ensure formatting.
- [x] I added docstrings to **any** new modules, classes, and methods.
- [ ] I updated the docstrings of **any** updated modules, classes, and methods.
- [x] I added/updated Python typings for any new/updated class and module.
- [ ] I updated the bot version (if necessary).
- [x] I ran mypy and reviewed any shown errors related to my changes.

# Important Info

- [ ] This pull request adds new `pip` dependencies.
- [ ] This pull request adds new external dependencies.
- [ ] This pull request changes the required versions of some dependencies.

# Issues Closed
My pull request closes the following issues:
N/A

# Thank You
Thank you for your contribution to Scioly.org! This pull request will be reviewed
in a promptly manner. If not done, please feel free to contact @cbrxyz.

Nydauron committed:
Add top-level sync command and sync script

It is generally best practice to not rate limit the slash command
registration endpoint, so syncing should be done manually rather than on
ready. A script is provided to simply login the bot and sync commands in
case the sync command signature ever gets modified.

cbrxyz commented on bot.py:
```diff
@@ -334,6 +354,35 @@ async def listen_for_response(
 discord.utils.setup_logging(handler=handler)
 
 
+@bot.tree.command(
+    description="Syncs command list. Any new commands will be available for use.",
+)
+@app_commands.checks.has_any_role(ROLE_STAFF, ROLE_VIP)
+async def sync(interaction: discord.Interaction):
```
Instead of just having a new file that you have to run to sync the `/sync` command, just make this a normal command and not an app command. Like `!sync`. Then you can get rid of the other file.

Nydauron replied:
There is a check to avoid legacy commands in `on_message`, so it would require modifying the regex, looking at the argument, then passing it to be processed iff it begins with `"sync "`. Idk, it just seems a bit messy for something super basic + there being two separate definitions of where "sync" is referenced (one in the command function signature, one in `on_message()`).
 
If we still go this route, would we want to opt for a [hybrid command](https://discordpy.readthedocs.io/en/latest/ext/commands/api.html?highlight=commands%20bot#discord.ext.commands.hybrid_command) instead?

cbrxyz commented on bot.py:
````diff
@@ -334,6 +354,35 @@ async def listen_for_response(
+    top-level command to prevent disabled cogs from disabling sync
+    functionality.
+
+    Note: Any changes to this command's signature will require
+    `Pibot.sync_commands()` which can be done by running:
+    ```
+    python sync_commands.py
+    ```
+    within your venv.
````
Thus you can remove all of this as well

cbrxyz commented on bot.py:
```diff
@@ -321,6 +324,23 @@ async def listen_for_response(
+        self,
+    ) -> tuple[list[app_commands.AppCommand], dict[int, list[app_commands.AppCommand]]]:
+        logger.info("Beginning to sync commands ...")
+        global_cmds_synced = await self.tree.sync()
```
You might want to add some sort of flag to only sync global commands when desired. Just FYI, this operation is kind of slow since it has to update the bot's profile across the entire Discord interface (and usually unnecessary, since most app commands are in guilds only).

Nydauron replied:
Do you want `!sync` to have a bool argument for syncing only guild or all commands? I am having a rather hard time getting a bool through to the command.

Nydauron replied:
> Do you want !sync to have a bool argument for syncing only guild or all commands?

Forgot to comment on an update, but this has been resolved.

cbrxyz reviewed (changes requested):
Thanks!

Nydauron committed:
Removed sync_commands.py script

Syncing commands now should be done using the legacy command

Nydauron commented on bot.py:
```diff
@@ -332,6 +361,36 @@ async def listen_for_response(
 discord.utils.setup_logging(handler=handler)
 
 
+@bot.command(
+    name=SYNC_COMMAND_NAME,
+    description="Syncs command list. Any new commands will be available for use.",
+)
+@commands.check(is_staff_from_ctx)
+async def sync(ctx: commands.Context, only_guild_commands: bool = False):
```
My only concern is that the boolean argument has an ambiguous meaning when typing it out in Discord (since there doesn't seem to be a way to display help info for a legacy command?).

I feel it would be better instead for this argument to be a string and to to check for `"all" | "guild" | "guildonly" | "guild-only"` and prompt an error msg stating command usage if the user puts something different.

Nydauron committed:
Added error handler and help text for sync command

cbrxyz commented on bot.py:
```diff
@@ -332,6 +361,56 @@ async def listen_for_response(
+        await ctx.send(res_msg)
+
+
+@sync.error
+async def sync_error(ctx: commands.Context, err):
+    logging.error(err)
+    if isinstance(err, commands.BadLiteralArgument):
+        literals = list(err.literals)
+        literals[-1] = f"or {literals[-1]}"
+        for literal in literals:
+            literal = f'"{literal}"'
+        assert sync.help is not None
+        msg = sync.help.format((", " if len(literals) > 2 else " ").join(literals))
+        return await ctx.send(msg, ephemeral=True)
+    await ctx.send(f"An unknown error occurred when syncing: {err}")
```
I really don't think all of this is super necessary, considering that this will only be run by admins rarely who can see the raw `BadLiteralArgument` exception in the console. But feel free to keep it if you want, doesn't hurt

Nydauron replied:
I just think it's better to provide feedback to the caller immediately rather than having to rely on having to look at the prod logs. Like you said, there's no real downside.

cbrxyz commented on bot.py:
```diff
@@ -332,6 +361,56 @@ async def listen_for_response(
+        for literal in literals:
+            literal = f'"{literal}"'
+        assert sync.help is not None
+        msg = sync.help.format((", " if len(literals) > 2 else " ").join(literals))
```
`", ".join(literals)`?

Nydauron replied:
Constructs the parallel sentence by joining the elements with `", "` (e.g. `", ".join(['"all"', '"guild"', 'or "guild-only"']) == '"all", "guild", or "guild-only"'`).

cbrxyz commented on bot.py:
```diff
@@ -332,6 +361,56 @@ async def listen_for_response(
+            literal = f'"{literal}"'
+        assert sync.help is not None
+        msg = sync.help.format((", " if len(literals) > 2 else " ").join(literals))
+        return await ctx.send(msg, ephemeral=True)
```
`ephemeral` does not do anything here because this message was not sent in an interaction

> [_ephemeral_ (**bool**) – Indicates if the message should only be visible to the user who started the interaction. If a view is sent with an ephemeral message and it has no timeout set then the timeout is set to 15 minutes. This is only applicable in contexts with an interaction.](https://discordpy.readthedocs.io/en/latest/ext/commands/api.html?highlight=badliteralargument#discord.ext.commands.Context.send)

Nydauron replied:
Yeah, I also realized that with:
https://github.com/cbrxyz/pi-bot/blob/58ccbe4d6c1024be92187d478b356049430359a9/bot.py#L381
that also doesn't work for the same reason. Shame it doesn't given it's part of the `ext.command.Command` function signatures, but oh well.

cbrxyz reviewed (changes requested):
Thanks!

Nydauron commented on bot.py:
```diff
@@ -37,6 +39,7 @@
 
 intents = discord.Intents.all()
 logger = logging.getLogger(__name__)
+SYNC_COMMAND_NAME = "sync"
```
> I don't think we'll change the name of the sync command.

True, but we need the sync name to prevent showing the "please use slash command" message (looking up the references for `SYNC_COMMAND_NAME` shows the two locations where it is read, compared to having two magic strings in either location).

Nydauron committed:
Removed separate sync command name variable

Nydauron committed:
Fixed prefix in command regex to be contructed from `BOT_PREFIX`

Nydauron committed:
Pins discordpy to earliest known good version of discordpy used

Nydauron committed:
Fixes pymongo dependencies due to conflicting and breaking API changes

See https://jira.mongodb.org/browse/PYTHON-4771 and
https://stackoverflow.com/questions/78999040/mongodb-importerror-with-new-pymongo-release
for a more in-depth reason.

Nydauron committed:
Add image build and push job to CI

This job is intended to only run on pushes to `master` or by manual
dispatch. Pull requests should not trigger this job and it should
instead be skipped automatically.

Nydauron committed:
Update action versions to consistent and non-deprecated versions

Certain versions of actions/cache v3 and v4 were deprecated at the end
of Jan 2025.

Nydauron committed:
Fixes insufficient scope error when pushing image from CI

Nydauron committed:
Replace hardcoded repo owner with CI variable

Nydauron committed:
Fixes conditional statement for build CI job

Nydauron committed:
chore: Add beanie ODM (#1)

* Added Beanie ODM initalizations

* Migrated cron collection to use ODM

* Created models.py file

* Changed which fields are indexable in `Cron` model

* Refactored remaining refs to cron collection DB operations

* Migrated ping collection to ODM

* Migrated tag collection to ODM

* Migrated invitational collection to ODM

* Migrated event collection to ODM

* Reimplemented business logic for `/event remove`

Replaces codepath for deleting partially deleted events (i.e. events
that do not exist in the database but have a server role) with a message
saying to delete the orphan server role manually. The reason why we
choose to *not* delete the role via the bot is because the bot can
potentially delete other roles not related to any event by accident.

* Migrated censor collection to ODM

* Migrated settings collection to ODM

* Remove `REPORTS` global

This was an unused global that used to be used for keeping track of
report discord message IDs. Users could submit reports using `/report`
and staff members could react to report messages to take action.

See commits for the last read/write references to the global:
- d61750b61d8091b82ffdaef87d01fa92237b8e94
- 191454883e1f319f5651303a2e7b9815ab62bf7c
- 1f4c01e09fb067e673f7e3901b26f7955716c4aa

* Remove redundant assert statements

* Remove mongo database wrapper class

At this point, as no functions within the class were being used, so it
is safe to remove. From now on, all database interactions should use
Beanie.

* Remove `get_reports()` and unused helper functions

* Add migrations folder

Nydauron committed:
fix: Correct typo preventing from legacy word ping from being removed (#3)

Nydauron committed:
Remove magic number used for discord options list (#2)

Nydauron committed:
chore: Change cbrxyz/pi-bot links to point to scioly/pi-bot

Nydauron committed:
chore: Change wiki links to use short URL format
